### DISCUSSION TITLE
Flow field visualisation for particular flow conditions in v4

### DISCUSSION THREAD

**Discussion by fwasilcz** [NatLabRockies/floris#892] [Q&A] [answered]
Hi, 
In v3 the calculate_horizontal_plane function allowed for specification of ws, wd and yaw angles.
In v4, the calculate_horizontal_plane function takes in the geometry parameters of the plane and findex_for_viz, which is supposed to be an int. How should I undestand this argument? Can someone explain or point me to some resource where I can read more about it?
I also tried using the method on fmodel with yaw implemented using set function, but it still produced the flow field for non-yawed conditions. 
Regards
Filip

**Comment by misi9170** [comment, marked as answer]
Hi @fwasilcz ,

You can find documentation [here](https://nrel.github.io/floris/_autosummary/floris.flow_visualization.html#floris.flow_visualization.calculate_horizontal_plane_with_turbines), but this may not entirely answer your question. You can also look at #868, which is where the change was introduced---I've also added a few details there. 

However, to summarize these (and they aren't the most clear, so certainly a summary is warranted!):
- In FLORIS v4, the user must `set` the wind conditions _prior_ to the call to `calculate_horizontal_plane`.
- If only a single condition is set (i.e. one wind speed, one wind direction, one turbulence intensity, and one "row" of yaw angles (i.e. one yaw angle for each turbine)), then `calculate_horizontal_plane()` can be called without specifying `findex_for_viz`, and the single condition that was set will be used. 
- The "findex" refers to the condition index
- If multiple conditions are set (i.e. multiple wind speeds, wind directions, etc), then the user may specify which to visualize in the call to `calculate_horizontal_plane()`. To specify the "first" condition, use `findex_for_viz=0`. This is also the default behavior if `findex_for_viz` is not passed. To specify the _n_ th condition, use `findex_for_viz=n`.

You can also see this usage in [examples/002_visualizations.py](https://github.com/NREL/floris/blob/main/examples/002_visualizations.py) and the scripts in [examples/examples_visualizations](https://github.com/NREL/floris/tree/main/examples/examples_visualizations). These show the `set` being called for the desired wind condition, followed by the `calculate_horizontal_plane()` call. However, they do not demonstrate that you may set multiple conditions, and then select the one you wish to visualize by setting `findex_for_viz`, as you asked about.

**Reply by misi9170** [reply]
Here is an example script that utilizes the `findex_for_viz` argument. In making this based on [examples/002_visualizations.py](https://github.com/NREL/floris/blob/main/examples/002_visualizations.py), I noticed that there is an incorrect comment in that example stating that the user _must_ specify a single condition with `set` in order to run visualization routines. I will fix that in an upcoming pull request.

```python
import matplotlib.pyplot as plt

import floris.layout_visualization as layoutviz
from floris import FlorisModel
from floris.flow_visualization import visualize_cut_plane


fmodel = FlorisModel("inputs/gch.yaml")

# Set the farm layout to have 8 turbines irregularly placed
layout_x = [0, 500, 0, 128, 1000, 900, 1500, 1250]
layout_y = [0, 300, 750, 1400, 0, 567, 888, 1450]
fmodel.set(layout_x=layout_x, layout_y=layout_y)

# For flow visualizations, the FlorisModel *may* be set to a single condition.
# Alternatively, multiple conditions may be set, if findex_for_viz is specified
# in the calculate_*_plane method.
fmodel.set(
  wind_speeds=[8.0, 8.0],
  wind_directions=[290.0, 200.0],
  turbulence_intensities=[0.06, 0.06]
) # Setting to findices (conditions)

horizontal_plane = fmodel.calculate_horizontal_plane(
    x_resolution=200,
    y_resolution=100,
    height=90.0,
    findex_for_viz=0, # Compute for the 0 findex (8.0 m/s, 290.0 deg, 0.06%)
)

# Plot the flow field with rotors
fig, ax = plt.subplots()
visualize_cut_plane(
    horizontal_plane,
    ax=ax,
    label_contours=False,
    title="Horizontal Flow with Turbine Rotors and labels (findex_for_viz=0)",
)


horizontal_plane = fmodel.calculate_horizontal_plane(
    x_resolution=200,
    y_resolution=100,
    height=90.0,
    findex_for_viz=1, # Compute for the 1 findex (8.0 m/s, 200.0 deg, 0.06%)
)

# Plot the flow field with rotors
fig, ax = plt.subplots()
visualize_cut_plane(
    horizontal_plane,
    ax=ax,
    label_contours=False,
    title="Horizontal Flow with Turbine Rotors and labels (findex_for_viz=1)",
)

plt.show()
```

running this produces a figure for each of the two conditions:
![horizontal_flow_with_rotors_and_labels_findex_0](https://github.com/NREL/floris/assets/39596329/00b182e8-8500-4e44-8e84-e01107f20740)
![horizontal_flow_with_rotors_and_labels_findex_1](https://github.com/NREL/floris/assets/39596329/c0af23c8-aaa6-43cb-b7d6-4fcb5878e122)

I did notice that there is a small "bug" in that, if you would like to also add turbine rotors to the image, `plot_turbine_rotors` uses the yaw angles from the 0th findex by default. The best approach here is to pass `yaw_angles` to `plot_turbine_rotors` (after a possible bit of manipulation).

**Comment by fwasilcz** [comment]
Thank you very much for your answer! Setting the conditions before calculating the plane worked very well.
Since I have a simulation with the entire windrose the findex_for_viz would be very painful to manage.

**Reply by misi9170** [reply]
You're right, working out the `findex_for_viz` would not be straightforward with a wind rose. Perhaps we will build some helper functions for that in future. In the meantime, I'm glad that setting the conditions beforehand is working as you need.
